### DISCUSSION TITLE
`family` modifier forces itself on all other providers using it - how to propagate the parameter?

### DISCUSSION THREAD

**Discussion by wujek-srujek** [rrousselGit/riverpod#573] [Q&A]
I have a `family` provider which takes a dynamic parameter. This parameter is defined in one page of the app, and then a new page is shown and receives the, and it is used to `read`/`watch` a provider, let's call it A.

Now, this provider A is used by another provider B, so B needs to be able to `watch` A, but it needs the parameter to do so (the code can be found here: https://github.com/rrousselGit/river_pod/discussions/564). As it goes deeper, more and more providers will need to be made `family` providers if the use A or B.

In the widget tree, some widgets need A, some need B, some may need C or D (supposing C uses B and. D used C), and they ALL needs to know the original parameter for A to be able to look up the provider they want. Essentially, it seems that once a provider is a `family` provider, it forces all other 'descending' providers using it (either directly, like B uses A, or indirectly, like D uses C uses B uses A) to be `family` as well. As a result, many widgets in the widget tree need to know the parameter in order to be able to look up the provider they want to use. What is the best way to pass this value to these widgets?

The one solution that I know of would be to simply pass the parameter to the widgets using a constructor argument, even if they don't need it directly, but their children do. But this is exactly one of the things that the original `provider` package tried to solve, i.e. such code would be a step backwards.

How is a use case like this best solved using Riverpod? Is it possible to somehow put a dynamic (defined at runtime) value into the provider container, and then all providers will be able to look it up themselves, without them being a `family`? I looked through the docs and the examples and couldn't find anything.

**Comment by rrousselGit** [comment]
The fact that family tend to be contagious is expected and natural. There is no reason to fight against it

But it is influenced by your architecture. Depending on how you model your state, you may need more or less of those. It's hard to evaluate outside of a case-by-case basis

**Reply by wujek-srujek** [reply]
My question is not to argue that `family` is wrong and it is indeed natural that it is contagious, I get that. The issue is my own code, which starts resembling code that I would write before I started using the previous `provider` package, and my inability to find a solution on my own with Riverpod.

Below is a sample app which works (Flutter 2.2.1, Riverpod 0.14.0+3). It doesn't do anything useful but it should present what I have in mind with my question. Please note that this is not actually my app and certain things are written the way they are only to present my doubts (e.g. the `IntermadiateWidget`).

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return const ProviderScope(
      child: MaterialApp(
        home: TimerPage(interval: Duration(minutes: 1)),
      ),
    );
  }
}

class TimerPage extends StatelessWidget {
  final Duration interval;

  const TimerPage({required this.interval});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Expanded(
            child: CountdownWidget(interval: interval),
          ),
        ],
      ),
      floatingActionButtonLocation: FloatingActionButtonLocation.startFloat,
      floatingActionButton: IntermediateWidget(interval: interval),
    );
  }
}

class CountdownWidget extends ConsumerWidget {
  final Duration interval;

  const CountdownWidget({required this.interval});

  @override
  Widget build(BuildContext context, ScopedReader watch) {
    final timerState = watch(
      timerNotifierProvider(interval),
    );

    // In reality, this widget also needs the status.

    return Center(
      child: Text('${timerState.remaining}'),
    );
  }
}

class IntermediateWidget extends StatelessWidget {
  final Duration interval;

  const IntermediateWidget({required this.interval});

  @override
  Widget build(BuildContext context) {
    return StopButton(interval: interval);
  }
}

class StopButton extends ConsumerWidget {
  final Duration interval;

  const StopButton({required this.interval});

  @override
  Widget build(BuildContext context, ScopedReader watch) {
    final timerStatus = watch(
      timerStatusProvider(interval),
    );

    return FloatingActionButton(
      backgroundColor: timerStatus == TimerStatus.completed
          ? null
          : Theme.of(context).colorScheme.error,
      onPressed: () {
        // Nothing to pop in this snippet, but in the real thing there is.
        // Navigator.of(context).pop();
        print('>>> $interval');
      },
      child: const Icon(Icons.stop_rounded),
    );
  }
}

// Riverpod.

enum TimerStatus {
  running,
  paused,
  // Stopped preemptively by the user.
  stopped,
  // Stopped naturally, i.e. it run to the end.
  completed,
}

@immutable
class TimerState {
  final Duration remaining;
  final TimerStatus status;

  const TimerState(
    this.remaining,
    this.status,
  );
}

class TimerNotifier extends StateNotifier<TimerState> {
  final Duration interval;

  TimerNotifier(this.interval)
      : super(TimerState(interval, TimerStatus.running));

  // start, stop logic etc.
}

final timerNotifierProvider =
    StateNotifierProvider.family<TimerNotifier, TimerState, Duration>(
  (ref, interval) => TimerNotifier(interval),
);

final timerStatusProvider = Provider.family<TimerStatus, Duration>(
  (ref, interval) => ref.watch(timerNotifierProvider(interval)).status,
);
```

- There are 3 widgets:
  1. `CountdownWidget`, which needs the whole `TimerState` and uses `timerNotifierProvider` to get the `state`. In the real app this provider is also used to get the `StateNotifier` and call methods like `start`, `stop` etc., that's why it uses `StateNotifier`.
  2. `StopButton`, which needs the `TimerStatus` and uses `timerStatusProvider` to get it.
  3. `IntermediateWidget` which doesn't do anything, but uses `StopButton` internally.
- As `timerNotifierProvider` is `family`, `timerStatusProvider` also needs to be `family` as it uses `timerNotifierProvider` internally and needs the parameter to be able to `watch` it.
- Both `CountdownWidget` and `StopButton`, as they call the providers, need to be passed the `interval` parameter.
- But also `IntermediateWidget` needs to know the `interval` as it needs to pass it down to `StopButton`; otherwise, this parameter is of no use to it.
- If another provider `providerC` had its own parameter and also used one of the providers above, i.e. need `interval` as well, any widgets that would need to look `providerC` up would need to pass both the `interval` and the new parameter. This way of coding gets less and less maintainable the deeper the provider dependency chain. (Granted, it probably won't be very deep.)

The question is about `interval` - in the current solution, all of the widgets must be passed the `interval` parameter explicitly. One of the reasons to use the old `provider` package was to prevent code like this one: I would register a `Provider.value` with `interval` at the top of the subtree of all widgets that needs it in any way, and then each widget below would be able to look the `interval` up and use it for `watch`/`read` etc. There would be no explicit passing of `interval` like in this code. As you may well know, the widget tree can be deeply nested and passing the value like this doesn't scale well.


How would you solve it using Riverpod? Or what would one improve the architecture in this case?

**Reply by wujek-srujek** [reply]
As I was writing it, an idea occurred to me: create a `StateProvider`, put the `interval` value into it, and have the other 2 providers use this `intervalProvider`, without them being a `family`. Here is the code (relevant parts only):
```dart

class TimerPage extends StatefulWidget {
  final Duration interval;

  const TimerPage({required this.interval});

  @override
  _TimerPageState createState() => _TimerPageState();
}

class _TimerPageState extends State<TimerPage> {
  @override
  void initState() {
    super.initState();

    context.read(intervalProvider).state = widget.interval;
  }

// ...

// Riverpod.

final intervalProvider = StateProvider<Duration?>(
  (ref) => null,
);

// ...

final timerNotifierProvider = StateNotifierProvider<TimerNotifier, TimerState>(
  (ref) {
    final interval = ref.watch(intervalProvider).state!;
    return TimerNotifier(interval);
  },
);

final timerStatusProvider = Provider<TimerStatus>(
  (ref) => ref.watch(timerNotifierProvider).status,
);
```
All the widgets not (except for the root page) don't need the `interval` parameter, and the providers don't need to be families any more. But there are issues: it compiles, but it fails at runtime with:
```
This UncontrolledProviderScope widget cannot be marked as needing to build because the framework is already in the process of building widgets.  A widget can be marked as needing to be built during the build phase only if one of its ancestors is currently building. This exception is allowed because the framework builds parent widgets before children, which means a dirty descendant will always be built. Otherwise, the framework might not visit this widget during this build phase.
The widget on which setState() or markNeedsBuild() was called was:
  UncontrolledProviderScope
The widget which was currently being built when the offending call was made was:
  Builder
...
```
leading back to the line:
```dart
class _TimerPageState extends State<TimerPage> {
  @override
  void initState() {
    super.initState();

    context.read(intervalProvider).state = widget.interval; // <= HERE
  }
```

Another issue is that now I need to deal with nullability as `intervalProvider` doesn't have any value initially.

**Reply by rrousselGit** [reply]
Maybe you are looking for ScopedProvider (or just all providers starting from 1.0.0-dev) to inject the family parameter deep in the tree

**Reply by wujek-srujek** [reply]
I tried this but this doesn't do what I need either, see here: https://github.com/rrousselGit/river_pod/issues/575.

**Reply by wujek-srujek** [reply]
For now I changed my code a bit - I use the idea with `StateProvider`, and once the interval is chosen by the user in a page, it is put into this provider; then, I navigate the app to the next page which has the other providers interacting with the 'interval provider'. 

It works fine and it did allow me to get rid of fall the families and made the code a bit simpler. The downside is that now I have a global 'dummy' provider as a workaround, which I wouldn't have otherwise (or at least I don't need right now, but maybe I would need to add it further down the road anyway).

I would still be very interested in how others deal with such use cases, e.g. @TimWhiting had similar reservations about families at one point (https://github.com/rrousselGit/river_pod/issues/6#issuecomment-657065126), I wonder how he implemented his feature in the end (I guess it was `ScopedProvider` + override, but this does't work for my case as I would need the override to be visible to depending providers).

**Reply by rrousselGit** [reply]
#575 is exactly how `package:provider` and all other InheritedWidget-based solutions works.
This may change in the future with code-generation, but for now it is your role to override a provider and all of its dependencies.

Which means:

```
ProviderScope(
  overrides: [
    iProvider.overrdeWithValue(i),
    wrapperProvider, // overrides the dependency
  ]
)
```

**Reply by wujek-srujek** [reply]
I understand now.  But just think about it - this way I need to hunt down all the providers that depend on the overridden one (directly or indirectly) and override them as well, by hand. This is not maintainable and very brittle. I can live with it, I will just need to keep in mind to be very careful when using this feature (if ever, maybe in tests?) as it doesn't do what I expect - maybe my coming from other platforms (like Java + Spring) and expecting this one to behave like them is skewing my perception.

**Reply by davidmartos96** [reply]
@rrousselGit You've mentioned code generation a couple times regarding this confusion around scoping. Would the generated code be equivalent to manually including other dependent providers as in your previous comment? 

I think its easier to see the scoping issues with provider than with riverpod. It may have to do with the magic around `watch`, people expect (myself included) the dependent providers would magically have the new value as well. :sweat_smile: 
Or maybe it has to do with the scoping being directly tied to the widget tree (in provider) vs a different relationships/chains layer (in riverpod)

**Reply by rrousselGit** [reply]
> Would the generated code be equivalent to manually including other dependent providers as in your previous comment?

Yes
